libs/base/Data/Rel.idr

```diff
@@ -0,0 +1,13 @@
+module Data.Rel
+
+import Data.Fun
+
+||| Build an n-ary relation type from a Vect of Types
+public export
+Rel : Vect n Type -> Type
+Rel ts = Fun ts Type
+
+public export
+liftRel : (ts : Vect n Type) -> (p : Rel ts) -> (Type -> Type) -> Type
```

Comment:
I think you ought to distinguish between `All : (ts : Vect n Type) -> Rel ts -> Type`
the function turning a `Rel` into a `Type` by universally quantifying over its domain
(its dual `Ex` would be useful too btw) and just using `map` on the return `Type` of
the relation.

Reply:
I don't know what `Ex` type would be, what are you thinking of?

Reply:
`Ex` would be the same type but a different function:

```idris
Ex : (ts : Vect n Type) -> (p : Rel ts) -> Type
Ex [] p = p
Ex (t :: ts) p = (x : t ** Ex ts (p x))
```

Reply:
How is it a dual of All? Because All produces product types and Ex sum types?

Reply:
Also, how would this be used?

Reply:
One interesting result is generalised (un)currying:

`Ex ts p -> t` ~ `All ts (map (\ pxs -> pxs -> t) p)`

That is to say:

`(x_1 : t_1 ** ... ** x_n : t_n ** p x_1 ... x_n) -> t`
is equivalent to
`(x_1 : t_1) -> ... -> (x_n : t_n) -> p x_1 ... x_n -> t`